### PR TITLE
npu Softmax op

### PR DESCRIPTION
**Pull request by OleNet**
### PR types

New features

### PR changes

OPs

### Describe

add op for softmax kernel [NPU]

image

image

### CODE REVIEW

**Comment by OleNet** [comment]
报错：
image

File: paddle/fluid/operators/softmax_op_npu.cc [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    tmp_out.mutable_data<T>(ctx.GetPlace());
+    framework::TensorCopy(
+        *out, ctx.GetPlace(),
+        ctx.template device_context<platform::DeviceContext>(), &tmp_out);
+    tmp_out.Resize(framework::make_ddim({first_dim, sec_dim}));
+
+    Tensor tmp_dOut(dOut->type());
+    tmp_dOut.Resize(framework::make_ddim({first_dim, sec_dim}));
+    tmp_dOut.mutable_data<T>(ctx.GetPlace());
+    framework::TensorCopy(
```
Comment (liym27) [inline paddle/fluid/operators/softmax_op_npu.cc]: tmp_out 和 tmp_dOut 都是对原数据进行拷贝的，目的是为了做 resize 吗？请用 ShareDataWith 替代数据拷贝。
示例：
```
Tensor tmp_out;
tmp_out.ShareDataWith(*out).Resize({first_dim, sec_dim});
```

File: python/paddle/fluid/tests/unittests/npu/test_softmax_op_npu.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            # 4 x 2
+            prob = fluid.layers.softmax(prediction, axis=1)
+
+            cost = fluid.layers.cross_entropy(input=prob, label=label)
+            loss = fluid.layers.mean(cost)
+            sgd = fluid.optimizer.SGD(learning_rate=0.01)
+            sgd.minimize(loss)
+
+        if run_npu:
+            place = paddle.NPUPlace(4)
```
Comment (liym27) [inline python/paddle/fluid/tests/unittests/npu/test_softmax_op_npu.py]: 用 `paddle.NPUPlace(0)`，因为之后 CI 测试不一定有卡4.

**Review by liym27 — APPROVED** [review]
LGTM
